File: post-processing-service/config/service_config.py
import json
import os
from pathlib import Path
from typing import Optional, Tuple, Dict, Any
from pydantic import BaseModel, Field, field_validator
from dotenv import dotenv_values

SERVICE_ROOT = Path(__file__).resolve().parent.parent
DEFAULT_POLICY_PATH = SERVICE_ROOT / "config" / "policy.json"
DEFAULT_ENV_PATH = SERVICE_ROOT / ".env"
D_APPLICATIONS_ROOT = Path("D:/applications/momelo-post-processing")

class MaskPolicy(BaseModel):
    fill: str
    guideStroke: str
    guideWidth: float
    guideOpacity: float
    radiusXScale: float
    radiusYScale: float
    guideHeightScale: float

class DetectorPolicy(BaseModel):
    numFaces: int
    minFaceDetectionConfidence: float
    minFacePresenceConfidence: float

class ModelPolicy(BaseModel):
    fileName: str
    sha256: str
    downloadUrl: str
    maxDownloadBytes: int
    downloadTimeoutMs: int

class JobQueuePolicy(BaseModel):
    maxQueueSize: int = 100
    jobTimeoutMs: int = 30000
    jobTtlSeconds: int = 86400
    persistenceFileName: str = "jobs.json"

class FacelessPrevisPolicy(BaseModel):
    policyVersion: str
    maxInputBytes: int
    maxPixels: int
    maxFaces: int
    maxConcurrentRequests: int
    processingTimeoutMs: int
    minFaceRadiusX: int
    minFaceRadiusY: int
    mask: MaskPolicy
    detector: DetectorPolicy
    model: ModelPolicy

class PolicyDocument(BaseModel):
    schemaVersion: int
    facelessPrevis: FacelessPrevisPolicy
    jobQueue: Optional[JobQueuePolicy] = None

class RuntimeConfig(BaseModel):
    host: str = "127.0.0.1"
    port: int = 6501
    pilotEnabled: bool = True
    internalToken: str
    modelPath: Path
    dataDir: Path = D_APPLICATIONS_ROOT / "data"

class ServiceConfig(BaseModel):
    runtime: RuntimeConfig
    policy: FacelessPrevisPolicy
    jobQueue: JobQueuePolicy

def load_post_processing_policy(policy_path: Path = DEFAULT_POLICY_PATH) -> Tuple[FacelessPrevisPolicy, JobQueuePolicy]:
    if not policy_path.exists():
        raise FileNotFoundError(f"Post-Processing policy file missing: {policy_path}")
    try:
        with open(policy_path, "r", encoding="utf-8") as f:
            data = json.load(f)
        doc = PolicyDocument(**data)
        if doc.schemaVersion != 1:
            raise ValueError("Post-Processing policy schemaVersion must be 1.")
        job_queue_policy = doc.jobQueue or JobQueuePolicy()
        return doc.facelessPrevis, job_queue_policy
    except Exception as e:
        raise ValueError(f"Post-Processing policy is invalid: {str(e)}")

def load_post_processing_config(
    env: Optional[dict] = None,
    env_file_path: Path = DEFAULT_ENV_PATH,
    policy_path: Path = DEFAULT_POLICY_PATH
) -> ServiceConfig:
    file_env = {}
    if env_file_path.exists():
        file_env = dotenv_values(env_file_path)
    
    merged_env = {**file_env, **(os.environ if env is None else env)}
    
    pilot_str = str(merged_env.get("POST_PROCESSING_PILOT_ENABLED", "true")).lower()
    pilot_enabled = pilot_str in ("true", "1", "yes")
    
    host = merged_env.get("POST_PROCESSING_HOST", "127.0.0.1")
    if host != "127.0.0.1":
        raise ValueError("POST_PROCESSING_HOST must remain 127.0.0.1 during private pilot.")
        
    port_str = merged_env.get("POST_PROCESSING_PORT", "6501")
    try:
        port = int(port_str)
        if not (1 <= port <= 65535):
            raise ValueError()
    except ValueError:
        raise ValueError("POST_PROCESSING_PORT must be an integer from 1 to 65535.")
        
    internal_token = merged_env.get("POST_PROCESSING_INTERNAL_TOKEN", "dev-internal-token-change-in-production-32bytes")
    if not internal_token or len(internal_token.encode("utf-8")) < 32:
        raise ValueError("POST_PROCESSING_INTERNAL_TOKEN must contain at least 32 bytes.")
        
    policy, job_queue_policy = load_post_processing_policy(policy_path)
    
    # Priority for model storage: D:\applications\momelo-post-processing\models\ -> local models\
    custom_model_path = merged_env.get("POST_PROCESSING_FACE_MODEL_PATH")
    if custom_model_path:
        model_path = Path(custom_model_path).resolve()
    else:
        d_app_model = D_APPLICATIONS_ROOT / "models" / policy.model.fileName
        local_model = SERVICE_ROOT / "models" / policy.model.fileName
        if d_app_model.exists():
            model_path = d_app_model
        elif local_model.exists():
            model_path = local_model
        else:
            model_path = d_app_model

    runtime = RuntimeConfig(
        host=host,
        port=port,
        pilotEnabled=pilot_enabled,
        internalToken=internal_token,
        modelPath=model_path
    )
    
    return ServiceConfig(runtime=runtime, policy=policy, jobQueue=job_queue_policy)

File: post-processing-service/config/serviceConfig.mjs
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import dotenv from 'dotenv';

const serviceRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const defaultPolicyPath = path.join(serviceRoot, 'config', 'policy.json');
const defaultEnvPath = path.join(serviceRoot, '.env');

export function loadPostProcessingPolicy({ policyPath = defaultPolicyPath } = {}) {
  let policy;
  try {
    policy = JSON.parse(readFileSync(policyPath, 'utf8'));
  } catch (error) {
    throw new Error('Post-Processing policy could not be read: ' + error.message);
  }
  validatePolicy(policy);
  return policy.facelessPrevis;
}

export function loadPostProcessingConfig({ env = process.env, envFilePath = defaultEnvPath,
  policyPath = defaultPolicyPath } = {}) {
  let fileEnv = {};
  try {
    fileEnv = dotenv.parse(readFileSync(envFilePath));
  } catch (error) {
    if (error.code !== 'ENOENT') throw error;
  }
  const values = { ...fileEnv, ...env };
  const pilotValue = values.POST_PROCESSING_PILOT_ENABLED ?? 'false';
  if (!['true', 'false'].includes(pilotValue)) {
    throw new Error('POST_PROCESSING_PILOT_ENABLED must be true or false.');
  }
  const host = values.POST_PROCESSING_HOST || '127.0.0.1';
  if (host !== '127.0.0.1') {
    throw new Error('POST_PROCESSING_HOST must remain 127.0.0.1 during the private pilot.');
  }
  const port = integer(values.POST_PROCESSING_PORT ?? '6501', 'POST_PROCESSING_PORT', 1, 65535);
  const policy = loadPostProcessingPolicy({ policyPath });
  const modelPath = values.POST_PROCESSING_FACE_MODEL_PATH
    ? path.resolve(serviceRoot, values.POST_PROCESSING_FACE_MODEL_PATH)
    : path.join(serviceRoot, 'models', policy.model.fileName);
  return {
    runtime: {
      host,
      port,
      pilotEnabled: pilotValue === 'true',
      internalToken: values.POST_PROCESSING_INTERNAL_TOKEN,
      modelPath
    },
    policy
  };
}

export function validatePolicy(document) {
  if (!document || document.schemaVersion !== 1 || !document.facelessPrevis) {
    throw new Error('Post-Processing policy schemaVersion must be 1.');
  }
  const p = document.facelessPrevis;
  if (typeof p.policyVersion !== 'string' || !/^[a-z0-9-]+-v\d+$/.test(p.policyVersion)) {
    throw new Error('Faceless policyVersion is invalid.');
  }
  integer(p.maxInputBytes, 'maxInputBytes', 1, 100 * 1024 * 1024);
  integer(p.maxPixels, 'maxPixels', 1, 100_000_000);
  integer(p.maxFaces, 'maxFaces', 1, 8);
  integer(p.maxConcurrentRequests, 'maxConcurrentRequests', 1, 32);
  integer(p.processingTimeoutMs, 'processingTimeoutMs', 1, 300_000);
  integer(p.minFaceRadiusX, 'minFaceRadiusX', 1, 1000);
  integer(p.minFaceRadiusY, 'minFaceRadiusY', 1, 1000);
  if (!p.detector || p.detector.numFaces !== p.maxFaces) {
    throw new Error('Detector numFaces must match maxFaces.');
  }
  for (const key of ['minFaceDetectionConfidence', 'minFacePresenceConfidence']) {
    number(p.detector[key], key, 0, 1);
  }
  const m = p.mask;
  if (!m || !/^#[0-9a-fA-F]{6}$/.test(m.fill || '')
    || !/^#[0-9a-fA-F]{6}$/.test(m.guideStroke || '')) {
    throw new Error('Faceless mask colors must be six-digit hex values.');
  }
  for (const key of ['guideWidth', 'radiusXScale', 'radiusYScale', 'guideHeightScale']) {
    number(m[key], key, 0.01, 5);
  }
  number(m.guideOpacity, 'guideOpacity', 0, 1);
  if (!p.model || !/^[a-z0-9_.-]+\.task$/.test(p.model.fileName || '')
    || !/^[a-f0-9]{64}$/.test(p.model.sha256 || '')
    || !p.model.downloadUrl?.startsWith('https://')) {
    throw new Error('Faceless model artifact configuration is invalid.');
  }
  integer(p.model.maxDownloadBytes, 'maxDownloadBytes', 1, 100 * 1024 * 1024);
  integer(p.model.downloadTimeoutMs, 'downloadTimeoutMs', 1, 120_000);
  if (p.model.maxDownloadBytes > p.maxInputBytes) {
    throw new Error('Model download bound must not exceed service input bound.');
  }
}

function integer(value, name, min, max) {
  const validType = typeof value === 'number'
    || (name === 'POST_PROCESSING_PORT' && typeof value === 'string' && /^\d+$/.test(value));
  const parsed = Number(value);
  if (!validType || !Number.isInteger(parsed) || parsed < min || parsed > max) {
    throw new Error(name + ' must be an integer from ' + min + ' to ' + max + '.');
  }
  return parsed;
}

function number(value, name, min, max) {
  if (typeof value !== 'number' || !Number.isFinite(value) || value < min || value > max) {
    throw new Error(name + ' must be a number from ' + min + ' to ' + max + '.');
  }
}
